redis/src/aio/connection.rs

```diff
@@ -256,108 +267,205 @@ where
 }
 
 /// Represents a `PubSub` connection.
-pub struct PubSub<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
-
-/// Represents a `Monitor` connection.
-pub struct Monitor<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
+pub struct PubSub<C = Pin<Box<dyn AsyncStream + Send + Sync>>> {
+    framed: Framed<C, ValueCodec>,
+    queue: VecDeque<Msg>,
+}
 
 impl<C> PubSub<C>
 where
     C: Unpin + AsyncRead + AsyncWrite + Send,
 {
-    fn new(con: Connection<C>) -> Self {
-        Self(con)
+    fn new(con: Connection<C>) -> PubSub<C> {
+        let framed = ValueCodec::default().framed(con.con);
+
+        PubSub {
+            framed,
+            queue: Default::default(),
+        }
+    }
+
+    // Deliver the PUBSUB command to this PUBSUB connection
+    async fn command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisResult<()> {
+        let mut out = Vec::new();
+
+        cmd.write_packed_command(&mut out);
+        self.framed.send(out).await?;
+
+        self.next_response().await
+    }
+
+    async fn next_response(&mut self) -> RedisResult<()> {
+        while let Some(item) = self.framed.next().await {
+            match item {
+                Ok(Ok(value)) => match Msg::from_value(&value) {
+                    Some(msg) => self.queue.push_front(msg),
+                    _ => return Ok(()),
+                },
+                _ => {}
+            }
+        }
+
+        Ok(())
+    }
+
+    /// placeholder
+    pub async fn next_message(&mut self) -> RedisResult<Msg> {
+        match self.queue.pop_back() {
+            Some(msg) => Ok(msg),
+            _ => match self.framed.next().await {
+                Some(Ok(Ok(value))) => match Msg::from_value(&value) {
+                    Some(msg) => Ok(msg),
+                    _ => Err(RedisError::from((
```

Comment:
but this could also catch a valid sub/unsub response, no?
This doesn't mean that the connection was necessarily dropped.